### PR TITLE
Secp256k1 support submodule

### PR DESCRIPTION
**Pull request by Zett98**
## Depends
- [x] #219
- [x] #221
## Problem

We don't have a Secp256k1 module to implement support for interop with tz2 addresses
## Solution

Add a Secp256k1 submodule to crypto module in tezos_interop.
## Related

- #199

### CODE REVIEW

File: tezos_interop/crypto.re [hunk 1]
```diff
@@ -232,3 +232,99 @@ module P256 = {
     let of_string = string => Base58.simple_decode(~prefix, ~of_raw, string);
   };
 };
+module Secp256k1 = {
+  open Libsecp256k1.External;
+  module K = Key;
+  let context = {
+    Mirage_crypto_rng_unix.initialize(); // TODO: this is probably a very bad thing to do
```
[the remark this answers is not in the document]
Comment (Zett98) [reply tezos_interop/crypto.re:239]: It fails during esy test, this fixes the issue, still this is bad

[the remark this answers is not in the document]
Comment (Zett98) [reply tezos_interop/crypto.re]: still fails :/

**Comment by Zett98** [comment]
> 
> 
> please clean the commit history for this PR.

done.

File: tezos_interop/crypto.re [hunk 2]
```diff
@@ -100,3 +100,101 @@ module Ed25519 = {
     let of_string = string => Base58.simple_decode(~prefix, ~of_raw, string);
   };
 };
+module Secp256k1 = {
+  open Libsecp256k1.External;
+  module K = Key;
+  let context = {
+    Mirage_crypto_rng_unix.initialize(); //TODO: adding this to to test files doesnt help ??
```
Comment (EduardoRFS) [inline tezos_interop/crypto.re:107]: it's probably because of the order of top level execution, it is depth first

File: tezos_interop/crypto.re [hunk 3]
```diff
@@ -100,3 +100,101 @@ module Ed25519 = {
...
+    let c = Context.create();
+    let rand_value =
+      Mirage_crypto_rng.generate(32) |> Cstruct.to_bytes |> Bigstring.of_bytes;
+    let randomized = Context.randomize(c, rand_value);
+    randomized
+      ? c : failwith("Secp256k1 context randomization failed. Aborting.");
+  };
```
Comment (EduardoRFS) [inline tezos_interop/crypto.re:114]: Can you explain to me the job of this global context? And also, how is it done in Tezos?

File: tezos_interop/crypto.re [hunk 4]
```diff
@@ -100,3 +100,101 @@ module Ed25519 = {
...
+
+  module Key = {
+    type t = K.t(K.public);
+    let equal = (a, b) => {
+      Bigstring.equal(K.to_bytes(context, a), K.to_bytes(context, b));
```
Comment (EduardoRFS) [inline tezos_interop/crypto.re:119]: Key.equal exists already

File: tezos_interop/crypto.re [hunk 5]
```diff
@@ -100,3 +100,101 @@ module Ed25519 = {
...
+    let to_raw = BLAKE2B_20.to_raw_string;
+    let of_raw = BLAKE2B_20.of_raw_string;
+    let to_string = t => Base58.simple_encode(~prefix, ~to_raw, t);
+    let of_string = string => Base58.simple_decode(~prefix, ~of_raw, string);
+  };
+  module Secret = {
+    type t = K.t(K.secret);
+
+    let equal = (a, b) => {
+      Bigstring.equal(K.to_bytes(context, a), K.to_bytes(context, b));
```
Comment (EduardoRFS) [inline tezos_interop/crypto.re:159]: Key.equal already exists

File: tezos_interop/crypto.re [hunk 6]
```diff
@@ -100,3 +100,101 @@ module Ed25519 = {
...
+    let of_raw = string => {
+      K.read_sk(context, string |> Bigstring.of_string) |> Result.to_option;
+    };
+
+    let to_string = t => Base58.simple_encode(~prefix, ~to_raw, t);
+    let of_string = string => Base58.simple_decode(~prefix, ~of_raw, string);
+  };
+  module Signature = {
+    [@deriving eq]
+    type t = string;
```
Comment (EduardoRFS) [inline tezos_interop/crypto.re:173]: I believe you should be using Sign.t here

File: tezos_interop/crypto.re (hunk 4 above)
Comment (Zett98) [reply tezos_interop/crypto.re]: changed

File: tezos_interop/crypto.re [hunk 7]
```diff
@@ -100,3 +100,101 @@ module Ed25519 = {
...
+    type t = K.t(K.public);
+    let equal = (a, b) => {
+      Bigstring.equal(K.to_bytes(context, a), K.to_bytes(context, b));
+    };
+    let size = K.compressed_pk_bytes;
+    let prefix = Base58.Prefix.secp256k1_public_key;
+    let encoding = {
+      let to_bytes = t => K.to_bytes(context, t) |> Bigstring.to_bytes;
+      let of_bytes_exn = b =>
+        Key.read_pk_exn(context, Bigstring.of_bytes(b));
```
[the remark this answers is not in the document]
Comment (Zett98) [reply tezos_interop/crypto.re:126]: yes, https://gitlab.com/tezos/tezos/-/blob/master/src/lib_crypto/secp256k1.ml#L63

**Pull request closed without merging**